src/cache.h

```diff
@@ -0,0 +1,46 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#pragma once
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+typedef enum { LRU_list, LFU_list, LRU_ghost_list, LFU_ghost_list } list_type_t;
+
+struct cache;
+
+/** cache_create - crate a new cache
+ * @return: a pointer points to new cache
+ */
+struct cache *cache_create();
```

Comment:
How can we specify the capacity?

Reply:
I make the user specify the capacity, but this implementation significantly reduces performance.

Reply:
Why?

Reply:
I guess the reason is that the hash function, which calculating by cache size, is called frequently. So, if we use a macro to represent size, the compiler can optimize it, possibly through constant propagation.

Reply:
Good point. Clarify your assumptions.

Reply:
I examine the assembly code, if we use a macro to represent size, the hash function only needs 5 instructions. But if we use a   variable to represent size, the hash function needs 9 instructions, including the dividing instruction which has large overhead.
* MACRO
```=
imul   $0x61c88647,%esi,%eax
mov    0x30(%rcx),%rdx
mov    (%rdx),%rdx
shr    $0x18,%eax
mov    (%rdx,%rax,8),%rax
```
* variable 
```=
imul   $0x61c88647,%esi,%eax
mov    $0x20,%ecx
xor    %edx,%edx
mov    0x30(%rdi),%r9
sub    0x6a84(%rip),%ecx        # 1204c <cache_size_bits>
shr    %cl,%eax
divl   0x6a80(%rip)        # 12050 <cache_size>
mov    (%r9),%rax
mov    (%rax,%rdx,8),%rax
```